crates/bevy_ecs/src/schedule/executor/multi_threaded.rs

```diff
@@ -522,132 +598,84 @@ impl MultiThreadedExecutor {
+                #[cfg(feature = "trace")]
+                let _span = system_span.enter();
                 // SAFETY:
                 // - The caller ensures that we have permission to
                 // access the world data used by the system.
                 // - `update_archetype_component_access` has been called.
-                unsafe { system.run_unsafe((), world) };
+                unsafe { system.run_unsafe((), context.world_cell) };
             }));
-            // tell the executor that the system finished
-            sender
-                .try_send(SystemResult {
-                    system_index,
-                    success: res.is_ok(),
-                })
-                .unwrap_or_else(|error| unreachable!("{}", error));
-            if let Err(payload) = res {
-                eprintln!("Encountered a panic in system `{}`!", &*system.name());
-                // set the payload to propagate the error
-                {
-                    let mut panic_payload = panic_payload.lock().unwrap();
-                    *panic_payload = Some(payload);
-                }
-            }
+            context.system_completed(system_index, res, system);
         };
 
-        #[cfg(feature = "trace")]
-        let task = task.instrument(
-            self.system_task_metadata[system_index]
-                .system_task_span
-                .clone(),
-        );
-
-        let system_meta = &self.system_task_metadata[system_index];
         self.active_access
             .extend(&system_meta.archetype_component_access);
 
         if system_meta.is_send {
-            scope.spawn(task);
+            context.scope.spawn(task);
         } else {
             self.local_thread_running = true;
-            scope.spawn_on_external(task);
+            context.scope.spawn_on_external(task);
         }
     }
 
     /// # Safety
     /// Caller must ensure no systems are currently borrowed.
-    unsafe fn spawn_exclusive_system_task<'scope>(
+    unsafe fn spawn_exclusive_system_task(
         &mut self,
-        scope: &Scope<'_, 'scope, ()>,
+        context: &Context<'_, '_>,
         system_index: usize,
-        systems: &'scope [SyncUnsafeCell<BoxedSystem>],
-        world: &'scope mut World,
     ) {
+        // SAFETY: `can_run` returned true for this system, which means
+        // that no other systems currently have access to the world.
+        let world = unsafe { context.world_cell.world_mut() };
         // SAFETY: this system is not running, no other reference exists
-        let system = unsafe { &mut *systems[system_index].get() };
+        let system = unsafe { &mut *context.systems[system_index].get() };
+        // Move the full context object into the new future.
+        let context = *context;
+        #[cfg(feature = "trace")]
+        let system_span = self.system_task_metadata[system_index]
+            .system_task_span
+            .clone();
```

Comment:
Is this clone necessary?

Reply:
I think so, because that gets sent to a spawned task, and the borrow of `system_task_metadata` won't live long enough.  I didn't change that; the previous code did the `clone()` in the call to `task.instrument` and I just moved the call here.

Reply:
OK sounds good to me.